### PR TITLE
Implement PublishResult Transaction

### PR DESCRIPTION
**Pull request by MSNTCS**
This PR should be merged after this [PR](https://github.com/CodeChain-io/foundry/pull/470) have merged.

### CODE REVIEW

File: voting/src/general_meeting.rs [hunk 1]
```diff
@@ -243,6 +243,21 @@ impl TxOwner for Context {
...
                 Ok(())
             }
+            PUBLISH_RESULT_TX_TYPE => {
+                let tx: PublishResultTransaction =
+                    serde_cbor::from_slice(&transaction.body()).map_err(|_| todo_fixthis)?;
+
+                let meeting_id = tx.action.meeting_id;
+                let meeting: GeneralMeeting = serde_cbor::from_slice(&self.storage().get(meeting_id.as_ref()).unwrap())
+                    .map_err(|_| ExecuteError::GeneralMeetingNotFound)
+                    .map_err(|_| todo_fixthis)?;
```
Comment (majecty) [inline voting/src/general_meeting.rs:253]:
```suggestion
                let meeting: GeneralMeeting = serde_cbor::from_slice(&self.storage().get(meeting_id.as_ref()).map_err(|_| ExecuteError::GeneralMeetingNotFound))
                    .unwrap("This code is serialized ......")
                    .map_err(|_| todo_fixthis)?;
```

**Comment by majecty** [comment]
@MSNTCS Could you rebase this on the #470?

**Comment by MSNTCS** [comment]
> @MSNTCS Could you rebase this on the #470?

I will do it after I merge #470.

**Comment by majecty** [comment]
@MSNTCS I want to review this PR after rebasing on the #470.
If not, I need to review it again after the rebase.

File: voting/src/general_meeting.rs [hunk 2]
```diff
@@ -119,7 +136,53 @@ impl Context {
...
+                let now = self.block_header.as_ref().unwrap().timestamp();
+                if meeting.tallying_time.time > now {
+                    return Err(ExecuteError::PublishTimeBeforeTallyingTime)
+                }
+                let mut final_result = Vec::new();
+
+                for _agenda in 0..meeting.number_of_agendas {
+                    let mut vote_result = VoteResult::new();
+                    for vote_id in &vote_box.votes {
+                        let vote = self.vote().get_vote(&vote_id).unwrap();
```
Comment (majecty) [inline voting/src/general_meeting.rs:165]: Please check the agenda number of the vote.
You can find it from the vote paper.

File: voting/src/general_meeting.rs [hunk 3]
```diff
@@ -243,6 +243,25 @@ impl TxOwner for Context {
                 }
                 Ok(())
             }
+            PUBLISH_RESULT_TX_TYPE => {
+                let tx: PublishResultTransaction =
+                    serde_cbor::from_slice(&transaction.body()).map_err(|_| todo_fixthis)?;
+
+                let meeting_id = tx.action.meeting_id;
+                let mut meeting: GeneralMeeting = {
```
Comment (majecty) [inline voting/src/general_meeting.rs:251]:
```suggestion
                let meeting: GeneralMeeting = {
```

File: voting/src/voting.rs [hunk 4]
```diff
@@ -285,6 +286,30 @@ impl VoteManager for Context {
 
         Ok(vote_paper.number_of_shares)
     }
+
+    fn get_number_of_agenda(&self, vote_id: &VoteId) -> Result<u32, Error> {
```
Comment (majecty) [inline voting/src/voting.rs:290]: I feel like `number of agenda` means the maximum number of agenda.
`agenda number` feels like the index number of the agenda.

How do you feel when you read the name?

If the name confuses us, we may use totally different names for them. Like `agenda_id`?

File: voting/src/voting.rs [hunk 5]
```diff
@@ -285,6 +286,30 @@ impl VoteManager for Context {
...
+        }
+        let vote: Vote = {
+            let bytes = &self.storage().get(vote_id.as_ref()).expect("Previously we checked the existence of the vote");
+            serde_cbor::from_slice(bytes).expect("Vote is serialized by this code")
+        };
+        let vote_paper_id = vote.corresponding_paper_id;
+
+        if !self.storage().has(vote_paper_id.as_ref()) {
+            return Err(Error::CorrespondingVotePaperNotFound)
+        }
```
Comment (majecty) [inline voting/src/voting.rs:302]: Let's remove this. If the vote exist, the voting paper must exist.

File: voting/src/voting.rs [hunk 6]
```diff
@@ -285,6 +286,30 @@ impl VoteManager for Context {
...
+        let vote_paper: VotePaper = {
+            let bytes = &self
+                .storage()
+                .get(vote_paper_id.as_ref())
+                .expect("Previously we checked the existence of the vote paper");
```
Comment (majecty) [inline voting/src/voting.rs:307]:
```suggestion
            let bytes = &self
                .storage()
                .get(vote_paper_id.as_ref())
                .expect("If the vote exist, the vote paper must exist");
```

File: voting/src/general_meeting.rs [hunk 7]
```diff
@@ -119,7 +136,56 @@ impl Context {
...
+                let meeting_id = tx.action.meeting_id;
+                let key = crate::generate_voting_box_key(&meeting_id);
+                let vote_box: VoteBox = {
+                    let bytes = &self.storage().get(key.as_slice()).ok_or_else(|| ExecuteError::VotingBoxNotFound)?;
+                    serde_cbor::from_slice(&bytes).expect("Vote box is serialized by this code")
+                };
 
+                let mut meeting: GeneralMeeting = {
+                    let bytes =
+                        self.storage().get(meeting_id.as_ref()).ok_or_else(|| ExecuteError::GeneralMeetingNotFound)?;
```
Comment (majecty) [inline voting/src/general_meeting.rs:152]:
```suggestion
                        self.storage().get(meeting_id.as_ref()).expect("If vote box exist, general meeting must exist")?;
```

File: voting/src/general_meeting.rs [hunk 8]
```diff
@@ -119,7 +136,56 @@ impl Context {
...
+                    return Err(ExecuteError::PublishTimeBeforeTallyingTime)
+                }
+                let mut final_result = Vec::new();
+
+                for agenda in 0..meeting.number_of_agendas {
+                    let mut vote_result = VoteResult::new();
+                    for vote_id in &vote_box.votes {
+                        let vote = self.vote().get_vote(&vote_id).unwrap();
+                        let vote_choice = vote.get_choice();
+                        let number_of_shares = self.vote().get_shares(&vote_id).unwrap();
```
Comment (majecty) [inline voting/src/general_meeting.rs:167]: Isn't it enough to use `shares`?
```suggestion
                        let shares = self.vote().get_shares(&vote_id).unwrap();
```

Naming is a difficult work. If you think other name is better, please telle me.

**Pull request merged**